LPD-62088 Strict virtual host access functionality doesn't work | Set CompanyThreadLocal in error dispatcher requests

jorgediaz-lr opened the pull request:
Hi @shuyangzhou 

I am sending this PR as Brian requested here https://github.com/brianchandotcom/liferay-portal/pull/166371#issuecomment-3371988681

In this PR I am fixing the `virtual.hosts.strict.access` as it is not working. I am also doing some additional changes to set the CompanyThreadLocal in error dispatcher requests, as it was detected problematic in LPD-60239.

**Root cause of `virtual.hosts.strict.access` problem**
The root cause of this issue is LPD-26723 because it added an additional call to `PortalInstances.getCompanyId(httpServletRequest);` in **SpSessionTerminationSamlPortalFilter** so the code that validates the valid virtual host at AbsoluteRedirectsFilter is no longer executed because once `PortalInstances.getCompanyId(httpServletRequest);` sets the request attributes and the validations are no longer executed in the following executions

**Solution**
To avoid this problem, I have moved the `PortalInstances.getCompanyId` call from AbsoluteRedirectsFilter  to a the **ValidHostNameFilter** filter that is always executed in first place.

Another idea that I implemented in this other PR https://github.com/brianchandotcom/liferay-portal/pull/165426 was to move this code to a new filter PortalInstancesFilter but this solution was rejected by Brian because he didn't want to add new filters.

**Setting CompanyThreadLocal in error dispatcher requests**

These changes also allows us to fix the https://liferay.atlassian.net/browse/LPD-60239 that was solved with a workaround: 
  - When an error page is rendered, any StatusLayoutUtilityPageEntryRequestContributor contributor needs to execute `PortalInstances.getCompanyId` because CompanyThreadLocal is not set.
  - I have configured the **ValidHostNameFilter** filter in liferay-web.xml file with both ERROR and REQUEST dispatchers, so the CompanyThreadLocal will be set for the error pages, avoiding future problems if other StatusLayoutUtilityPageEntryRequestContributor is implemented.
  - This is specially problematic for Saas customers with database partition enabled, because if CompanyThreadLocal is not set, some queries doesn't return the correct values
 
Let me know if you need any additional clarification

Thank you

liferay-continuous-integration commented:
To conserve resources, the PR Tester does not automatically run for every pull.

If your code changes were already tested in another pull, reference that pull in this pull so the test results can be analyzed.

If your pull was never tested, comment &quot;ci:test&quot; to run the PR Tester for this pull.

shuyangzhou commented:
@jorgediaz-lr There are multiple tickets involved in this pr. I will need to see this pr from a list of test cases point of view, rather than digging through the ticket history to find out related info.

Please prepare a list of detailed test cases for each involved ticket and put them here. I see SAML is involved, please make sure all steps are in actionable details, I don't know anything about saml, please don't omit any details.

After this, our goal will become passing all your test cases in the simplest way possible.

jorgediaz-lr commented:
Hi @shuyangzhou 

When you said:

> _Please prepare a list of detailed test cases for each involved ticket and put them here._ 

Do you mean just list the test cases here in a comment of the pull request or implement additional tests in the code?

If you also prefer, I think I can split this PR in two, the first one related to the "Strict virtual host access functionality doesn't work" and a second one related to "Set CompanyThreadLocal in error dispatcher requests" to make this easier to review it.

Thank you

shuyangzhou commented:
just test cases from related tickets, I didn't mean any new test cases.

No, don't split the pull, those problems are related, we need to deal with them as a whole.

but after a brief reading, is it true that if you revert LPD-26723 changes, all problems other than the problem in LPD-26723 are solved?

jorgediaz-lr commented:
If you revert [LPD-26723](https://liferay.atlassian.net/browse/LPD-26723) changes, will only solve the error about `virtual.hosts.strict.access` explained [LPD-62088](https://liferay.atlassian.net/browse/LPD-62088). 

The changes related to set CompanyThreadLocal in error dispatcher requests is unrelated to [LPD-26723](https://liferay.atlassian.net/browse/LPD-26723) but I included here because moving the `PortalInstances.getCompanyId` call to another Filter made easier to fix it (_this second part could be moved to other PR_)

shuyangzhou commented:
Ok, so it sounds like we are really dealing with 2 unrelated issues here?

Please group your commits to deal with the 2 issues separately (still in same pr), and tell me which part fixes which issues.
And the most important thing is the list of detailed test cases, we will be using them for acceptance checking.

jorgediaz-lr commented:
Hi @shuyangzhou 

I have updated the pull request, now you have two separate blocks:

- **Commits 1 to 5 =>** It fixes the problem related to `virtual.hosts.strict.access` as explained in [LPD-62088](https://liferay.atlassian.net/browse/LPD-62088)
- **Commits 6 to 9, with the "Set CompanyThreadlocal in error pages" prefix =>** It sets CompanyThreadLocal in error dispatcher requests to improve the fix of [LPD-60239](https://liferay.atlassian.net/browse/LPD-60239)

About the test cases for each ticket, here there are all the information that I have collected. Let me know if you need additional information:

### First ticket: [LPD-26723](https://liferay.atlassian.net/browse/LPD-26723) [this one causes the second ticket error]
_(solution already committed in master)_

When a user is signed out from the Identity Provider, the Liferay session is not correctly invalidated, so the user icon menu is shown instead of the Sign In link

**Purpose of SpSessionTerminationSamlPortalFilter**: Its function is to enforce SAML Single Logout (SLO) in Liferay. It acts as the final mechanism to ensure a user is fully logged out after their central session is ended by an Identity Provider (IdP).

The Scenario and the Filter's Action:
1. Central Logout: A user logs out from the central IdP or another connected application.
2. Liferay Notified: The IdP sends a logout request to Liferay, which marks the user's corresponding SAML session record (SamlSpSession in the DB) as "terminated=true" The user's local Liferay session may still be active.
3. The Interception: On the user's very next action (e.g., clicking a link), the SpSessionTerminationSamlPortalFilter intercepts the request.
4. Enforcement: SpSessionTerminationSamlPortalFilter checks the SamlSpSession, sees it's terminated, and immediately invalidates the users session from the httpServletRequest (https://github.com/liferay/liferay-portal/blob/5801d042a9ddec855197ce23690dad33540fae05/modules/dxp/apps/saml/saml-opensaml-integration/src/main/java/com/liferay/saml/opensaml/integration/internal/servlet/profile/BaseProfile.java#L484-L526) and redirects them to complete the logout process

5. Results:
    - **Expected Results:** User is completely signed-out of the SP, so the “Sign In” link is present.
    - **Actual Results before LPD-26723 was solved:** User is signed out part-way through the page rendering, so the user icon is present, although no other privileges are present.

**Additional notes:**

- Before [LPD-26723](https://liferay.atlassian.net/browse/LPD-26723) changes, this `SpSessionTerminationSamlPortalFilter` was executed after the `AbsoluteRedirectsFilter`, which initializes the `CompanyThreadLocal `with the current companyId based off the virtual host provided. 

- However, the `AbsoluteRedirectsFilter` is [re-initializing the session here](https://github.com/liferay/liferay-portal/blob/ba9d2ba803a68c8d511640bc697f5266cada47d8/portal-impl/src/com/liferay/portal/servlet/filters/absoluteredirects/AbsoluteRedirectsFilter.java#L75-L88) before the `SpSessionTerminationSamlPortalFilter` has a chance to properly invalidate it. This leads to certain user related data still being present and causes the reported error.

- To solve this in [LPD-26723](https://liferay.atlassian.net/browse/LPD-26723) changes, the developers from app-security team moved `SpSessionTerminationSamlPortalFilter` before the `AbsoluteRedirectsFilter` to be able to change the user's session before the other filter changes it. They also added a call to `PortalInstances.getCompanyId` to populate the `CompanyThreadLocal` as it is used by this SAML filter. (this new call caused the second bug)

### Second ticket: [LPD-62088](https://liferay.atlassian.net/browse/LPD-62088)
_(not fixed in master)_

If you activate the strict virtual host access functionality and you request a non-existing virtual host in the VirtualHost table, the expected behavior is Liferay returns an error page with the “Unknown Virtual Hostname” but the current behavior is the default instance is returned.

**Steps to reproduce:**

1. Set in the portal-ext.properties:
```
virtual.hosts.valid.hosts=*
virtual.hosts.strict.access=true
```
2. Configure several hostnames in your /etc/hosts for example:
```
127.0.0.1 example1.com
127.0.0.1 example2.com
127.0.0.1 example3.com
```
3. Create new instances using `example1.com` and `example2.com` hostnames
4. Check access to these hostnames => they should work :heavy_check_mark:
5. Check access to `example3.com` hostname:
    - **Expected behavior:** an error is produced and you cannot access to the example3.com hostname `Unknown Virtual Hostname: example03.com` :heavy_check_mark:

    - **Wrong behavior:** the default instance is displayed because no instance is configured with the example3.com hostname :x:

**Additional notes:**

 - The root cause of this error is that `PortalInstances.getCompanyId` is called in **AbsoluteRedirectsFilter** with the strict parameter = PropsValues.VIRTUAL_HOSTS_STRICT_ACCESS, so with `virtual.hosts.strict.access=true` we would throw/catch the exception and display the error, but after [LPD-26723](https://liferay.atlassian.net/browse/LPD-26723) changes a new call to `PortalInstances.getCompanyId` without the strict parameter is executed before in **SpSessionTerminationSamlPortalFilter**. 
 - This previous call causes that the call in AbsoluteRedirectsFilter doesn’t do any strict validation as the companyId and LayoutSet data is stored in the request and it is not calculated again in following calls to this `PortalInstances.getCompanyId` method.

 - To solve this problem, my idea is to remove all calls to `PortalInstances.getCompanyId` in the filters and leave only one with the   strict parameter = PropsValues.VIRTUAL_HOSTS_STRICT_ACCESS in the first filter. (_My initial idea was to create a new specific filter just to call `PortalInstances.getCompanyId`, but Brian didn’t liked it, so I moved it to ValidHostNameFilter that is the first one_)

### Third ticket [LPD-60239](https://liferay.atlassian.net/browse/LPD-60239)
_(workaround committed in master, only for CommonStatusLayoutUtilityPageEntryRequestContributor)_

If DB Partition is enabled and you request a non existing resource when the error pages are rendered, the CompanyThreadLocal is not set, so new contributors such as CommonStatusLayoutUtilityPageEntryRequestContributor has a NoSuchRoleException because without this ThreadLocal set, the queries to the database are executed in the wrong partition.

**Steps to reproduce:**

1. Configure Liferay with DB Partition enabled `database.partition.enabled=true`
2. Startup Liferay
3. Create a virtual instance: example1.com
4. Make sure your /etc/hosts file includes 127.0.0.1 example1.com
5. Navigate to a non-existent resource http://example1.com:8080/test.png
     - **Expected behavior:** The utility error page is shown and no errors are thrown in the logs
     - **Wrong behavior before LPD-60239 solution was committed:** The utility error page is shown but an NoSuchRoleException is thrown in the logs

**Additional notes:**

- In https://github.com/brianchandotcom/liferay-portal/pull/164514, as a workaround, the problem was solved adding a call to  `PortalInstances.getCompanyId` inside the contributor to set the CompanyThreadLocal, but this call shouldn’t be necessary and the CompanyThreadLocal should already be set in the error pages logic before the contributor is called. 
- This also will cause the same problem in future error pages contributors that we will implement. They won’t work correctly unless they also apply the same workaround of calling `PortalInstances.getCompanyId`
- To solve this problem, my idea is to change the configuration of ValidHostNameFilter in liferay-web.xml so the filter is executed for requests for both REQUEST and ERROR dispatcher and  `PortalInstances.getCompanyId` is also executed for the error pages

shuyangzhou commented on portal-impl/src/com/liferay/portal/servlet/filters/validhostname/ValidHostNameFilter.java:
```diff
@@ -5,27 +5,62 @@
 		String serverName = httpServletRequest.getServerName();
 
-		if (!PortalUtil.isValidPortalDomain(
+		if ((httpServletRequest.getDispatcherType() != DispatcherType.ERROR) &&
```
@jorgediaz-lr looking at the total diff of ValidHostNameFilter. I can see it starts to filter `ERROR` request too. but in case it is indeed serving an error request, this if condition `httpServletRequest.getDispatcherType() != DispatcherType.ERROR` is completely bypassing its original logic.

So functionally speaking when under error request serving this ValidHostNameFilter actually does not validate hostname at all, it is behaving as a total different filter.

This makes me feel like we are abusing ValidHostNameFilter just for its filterchain postion.

I think it will make more sense and make things more clear to create a dedicated filter for strict virual host checking, and put it in front of the chain. Leave the ValidHostNameFilter as it used to be (validating serverName) and just for `REQUEST`. the new filter handles both `ERROR`/`REQUEST` for strict virtual host checking and populate company thread local.

shuyangzhou commented:
See https://github.com/shuyangzhou/liferay-portal/pull/12075 

@jorgediaz-lr I just made a few small changes, please watch over for the CI results. As long as CI is clean, the new pull can be forwarded.

jorgediaz-lr commented:
Hi @shuyangzhou 

About your comment in this PR:: 
> You just need to deal with these 2 filters. After this pull, please start a thread with SF team to ask for a SF rule. "Any filter that extends BasePortalFilter must respect the portal properties configuration". For those BasePortalFilter subclass that overrides isFilterEnabled() they most likely should not be a BasePortalFilter, and the matching portal.properties line should be removed too.

I have created https://liferay.atlassian.net/browse/LPD-69645 ticket requesting it.
